### ISSUE TITLE
Automate without powershell using QR codes

### ISSUE THREAD

**Issue by Richard1793** [DrMikeyS/COVIDVaccinePatientSlips#20]
We have VDI terminals without powershell.

However, most forms can be navigated with keyboard shortcuts.  So I'm thinking a series of QR codes for those key presses could navigate the form?  In a nutshell it'd be the powershell commands written in QR code.  Where powershell has a pause to allow the page to load the next section we'd simple swipe the next QR once the page is ready.

Has anyone tried this?

**Comment by Richard1793** [comment]
I've gone down this route because I want to avoid installing extensions both due to lockdown nature of our terminals.

**Comment by DrMikeyS** [comment]
I have not tried this but it is entirely feasible. I have actually abandoned the powershell approach in our LVS because we have clinicians data entering and it was too complex and error prone. A QR approach you describe is much more safe (as each step is chosen by a person).

**Comment by Richard1793** [comment]
I've started on this.  You should be able to see it in my branch QR-Auotfill (yes there's a typo!)

The coding is chr09 to move to the next field and " " to select a button/checkbox.

To use the autofill sticker option:
1 - Click on sole practitioner  then scan pt id (ie the first QR code)
1a - If using the test xls you'll need to manually idenitfy a real test patient then click confirm
2 - After confirming PDS click on th patient name field (ie the one directly below the confirm/unlink button
3 - Then scan the second bar code and it will set consent to email to "no" and emergency contact to "no".  They often catch us out so even this first step is a help for data entry.

My next development steps would be:
1 - Read the vaccination setting from index.html and select the correct option
2 - Select resident in care home if setting is care home (note care home staff will need manual work correction)

At this point we could set ethnicity for "not stated", but that's favouring expediency over data collection and gives us rubbish data.  As it happens our demographic on vaccines to date is 97% white british so for us it'd make sense to default to white british and to keep going with autofill.  Users could manually correct by scrolling back up after the autofill.  That will be slow but minimal for us.  However this could cause problems for other practices with diverse ethnicities who'd need to do a lot of scrolling back up to correct this.

Perhaps offer 3 options for ethnicity in index.html
a - Select a default from radio /drop list
b - Select "Not stated" 
c - Finish autofill at ethnicity

BUT if this is a second vaccine we could pick up the  ethnicity used for the first dose in the pinnacle data file although that would be labourious to code and to prepare the input xls.  Too complex so park that for a later update?

Once past ethnicity I was planning to go up to and including the batch details.  This is my main target because it is prone to error if your site has several batch numbers in stock.

After that it'll be manual entry for left/right arm etc.

That's me done for now.  Drop a note in here if you are going to do some work on this so we don't duplicate efforts.

**Comment by Richard1793** [comment]
ps The main changes are:

index.html:
 - new page 2c  to gather the extra data (plus tweaks to get the page to show/hide)

function.js
 - a bit of new logic in the sticker routine to insert the relevant QR codes.  I've popped in some text markers (eg "boo") on the sticker while I was playing with the layout before I added the QR codes.  Just realised they are still then but they can be deleted now.
 - there's a comment "RH up to here" to show where the autofill QR is defined.  Makes it easier to pick up where I left off!

**Comment by DrMikeyS** [comment]
Great suggestions Richard. What I am thinking is could we keep the automation code separate from the patient ID code?

Two reasons for this:

1. Long strings lead to complex QR codes that need to be printed larger in order to render
2. This requires changes to the existing and ever complicating codebase

If we were to keep the current stickers the same, but have a second QR that would be static and does what you describe above, users could keep he static code printed on their desks. One scan for the patient data, one to autofill the rest. 

If you are happy with this, we could then create a separate HTML file and JS file for this new element of the project, which prevents bugs leaking into the other code.

Also, I can't see your fork. Have you pushed it to the public base?

**Comment by Richard1793** [comment]
I was in two minds about putting it on a seperate slip.  The main reason for having it on the sticker was because it will have the bacth number in the QR code, so seems to make sense to keep it pysically together to avoid the risk of using a bactch number on a slip that didn't match the sticker.

So far the QR code is short, visually it's a lot simpler than the id QR.

Sorry you can't see my branch.  It is showing on my fork under my profile, screenshot below.  Let me know if I need to push another button to make it visible.

The three updated files are attached as a manual workaround ;-)

![image](https://user-images.githubusercontent.com/80758997/113033347-d4ff1a00-9188-11eb-8a92-22adbfa47e53.png)

[210330 autofill QR.zip](https://github.com/DrMikeyS/COVIDVaccinePatientSlips/files/6231108/210330.autofill.QR.zip)

**Comment by DrMikeyS** [comment]
OK you have me just about sold. My remaining concerns are:

1. Is it worth the loss of the final ID check the spine lookup allows? When I vaccinate, I scan the ID details and say "So just to check, you are Mrs Jones of 10 High Street?" -  that address check has avoided in my case at least one incident where the patient had been given the wrong card by reception.
2. How does the system hold up if there is a time delay from the spine lookup?

I haven't actually tested the code as I am on leave and I don't have a scanner at home, but looking at your code it seems both problems are possibly solved with QR code 1 having the patient ID details and QR code 2 answers the subsequent questions.

Moving forward, there are a lot of people using the system nationally so I am wary of releasing big changes without robust testing. Therefore I propose (if you are happy) to create a public "beta" branch that we will publicise and try to get testers for. Once we are happy that the changes are stable and nothing breaks we will merge into the main "stable" branch. The beta branch would start with your code, once you are happy with it, and I will add you as a contributor who can edit the main repo.

**Comment by Richard1793** [comment]
That's why there are two steps.  

First QR takes the user from "Sole practioner" button to the confirm button.  Then manual click on"Confirm" with the usual verbal confirmation. ( already written in my branch also attached in prev post)

Second QR  (yet to do) takes the user from the Patient name box to the end.

So the 2 QR codes  force the user to confirm id and also allow for the delay in looking up PDS.

ps - We ended up with 5 Engleberts  in the space of 10 appointments (the actual name was as unusual as that), two Johns and two Ednas.   Bonkers.

**Comment by DrMikeyS** [comment]
Perfect. I have created a "dev" branch of the project that this development should be focused on.
